File: src/tonemap/LightnessToneMap.cpp
#include "java/lang/Math.h"
#include "common/error.h"
#include "common/cie.h"
#include "common/Statistics.h"
#include "tonemap/LightnessToneMap.h"

LightnessToneMap::LightnessToneMap() {
}

LightnessToneMap::~LightnessToneMap() {
}

void
LightnessToneMap::init() {
}

ColorRgb
LightnessToneMap::scaleForComputations(ColorRgb radiance) const {
    logWarning("ScaleForComputations", "%s %d not yet implemented", __FILE__, __LINE__);
    return radiance;
}

ColorRgb
LightnessToneMap::scaleForDisplay(ColorRgb radiance) const {
    float max = radiance.maximumComponent();
    if ( max < 1e-32 ) {
        return radiance;
    }

    // Multiply by WHITE EFFICACY to convert W/m^2sr to nits
    // (reference luminance is also in nits)
    float scaleFactor = lightness(WHITE_EFFICACY * max);
    if ( scaleFactor == 0.0 ) {
        return radiance;
    }

    radiance.scale(scaleFactor / max);
    return radiance;
}

float
LightnessToneMap::lightness(float luminance) {
    if ( GLOBAL_statistics.referenceLuminance == 0.0 ) {
        return 0.0f;
    }

    float relativeLuminance = luminance / (float)GLOBAL_statistics.referenceLuminance;
    if ( relativeLuminance > 0.008856 ) {
        return 1.16f * java::Math::pow(relativeLuminance, 0.33f) - 0.16f;
    } else {
        return 9.033f * relativeLuminance;
    }
}

File: java/src/main/java/vsdk/toolkit/tonemap/LightnessToneMap.java
package vsdk.toolkit.tonemap;

import vsdk.toolkit.common.Cie;
import vsdk.toolkit.common.ColorRgb;
import vsdk.toolkit.common.Error;
import vsdk.toolkit.common.statistics.Statistics;

public final class LightnessToneMap extends ToneMap {
    private static float lightness(float luminance) {
        if (Statistics.instance().radiance.referenceLuminance == 0.0) {
            return 0.0f;
        }

        float relativeLuminance = luminance / (float)Statistics.instance().radiance.referenceLuminance;
        if (relativeLuminance > 0.008856) {
            return 1.16f * (float)Math.pow(relativeLuminance, 0.33) - 0.16f;
        }
        else {
            return 9.033f * relativeLuminance;
        }
    }

    public LightnessToneMap() {
    }

    /*toneMapOptions*/
    @Override
    public void init(ToneMappingContext toneMapOptions) {
    }

    @Override
    public ColorRgb scaleForComputations(ColorRgb radiance) {
        Error.warning("ScaleForComputations", "%s %d not yet implemented", "LightnessToneMap.cpp", 0);
        return radiance;
    }

    @Override
    public ColorRgb scaleForDisplay(ColorRgb radiance) {
        float max = radiance.maximumComponent();
        if (max < 1e-32) {
            return radiance;
        }

        // Multiply by WHITE EFFICACY to convert W/m^2sr to nits
        // (reference luminance is also in nits)
        float scaleFactor = lightness(Cie.WHITE_EFFICACY * max);
        if (scaleFactor == 0.0) {
            return radiance;
        }

        radiance.scale(scaleFactor / max);
        return radiance;
    }
}
